[M3] 볼링 점수판(점수 계산) 미션 리뷰 요청

OhSeungCheol opened the pull request:
안녕하세요.

[M3] 볼링 점수판(점수 계산) 리뷰 부탁드립니다.

감사합니다.

Commit:
[M2] refactor: 불필요 함수 삭제

참고: https://github.com/next-step/java-bowling/pull/896

Commit:
[M2] refactor: UI 관련 로직 분리

참고: https://github.com/next-step/java-bowling/pull/896

Commit:
[M2] refactor: 어색한 키워드 수정 (Last > Latest)

참고: https://github.com/next-step/java-bowling/pull/896

Commit:
[M2] refactor: 코드 간결화

참고: https://github.com/next-step/java-bowling/pull/896

Commit:
[M2] refactor: 함수 순서 변경

참고: https://github.com/next-step/java-bowling/pull/896

Commit:
[M2] refactor: 테스트 코드 삭제

참고: https://github.com/next-step/java-bowling/pull/896

Commit:
[M2] refactor: 함수 순서 변경(2)

참고: https://github.com/next-step/java-bowling/pull/896

Commit:
[M3] docs: 기능명세서 작성

참고: https://edu.nextstep.camp/s/DVVUSLFe/ls/RmTfcj3f

Commit:
[M3] fix: 보너스 점수 유효성 판단 함수 동작 오류 수정

참고: https://edu.nextstep.camp/s/DVVUSLFe/ls/RmTfcj3f

Commit:
[M3] feat: 프레임의 일반 점수를 계산하는 함수 개발

참고: https://edu.nextstep.camp/s/DVVUSLFe/ls/RmTfcj3f

Commit:
[M3] test: 보너스 점수 조회 기능 테스트 추가

참고: https://edu.nextstep.camp/s/DVVUSLFe/ls/RmTfcj3f

Commit:
[M3] feat: 일반 프레임 보너스 점수 조회 기능 개발

참고: https://edu.nextstep.camp/s/DVVUSLFe/ls/RmTfcj3f

Commit:
[M3] refactor: 프레임 일반 점수 조회 함수 네이밍 변경 (getScore -> getPinScore)

참고: https://edu.nextstep.camp/s/DVVUSLFe/ls/RmTfcj3f

Commit:
[M3] feat: 결과 보너스로 마지막 프레임의 보너스 점수를 계산하는 함수 추가

참고: https://edu.nextstep.camp/s/DVVUSLFe/ls/RmTfcj3f

Commit:
[M3] feat: 누적 점수 계산 기능 개발

참고: https://edu.nextstep.camp/s/DVVUSLFe/ls/RmTfcj3f

Commit:
[M3] feat: 보너스 점수가 아직 준비가 안되어 있으면 예외를 발생시키도록 수정

참고: https://edu.nextstep.camp/s/DVVUSLFe/ls/RmTfcj3f

Commit:
[M3] test: NotReady 예외가 발생하는 케이스에 대한 테스트 수정

참고: https://edu.nextstep.camp/s/DVVUSLFe/ls/RmTfcj3f

Commit:
[M3] feat: 누적 점수 출력 기능 개발

참고: https://edu.nextstep.camp/s/DVVUSLFe/ls/RmTfcj3f

OhSeungCheol commented:
내부에서 사용하는 용어 중 헷갈릴만한 부분에 대해 따로 설명 추가합니다.

PinScore : 한 프레임에서 넘어뜨린 핀의 합
BonusScore : 스트라이크, 스페어일 경우에 발생하는 보너스 점수(다음 투구에서 넘어뜨린 핀의 개수) 값
FrameScore : PinScore + BonusScore
TotalScore : 이전 프레임들을 포함한 현재 프레임의 FrameScore 누적 값 

감사합니다.

lxxjn0 commented on src/main/java/bowling/domain/BowlingGame.java:
```diff
@@ -4,24 +4,13 @@ public class BowlingGame {
-            return true;
-        }
-        return false;
+        return records.getRecordCount() >= RuleConfig.NUMBER_OF_FRAME && records.isEndLastFrame();
```
`records.getRecordCount() >= RuleConfig.NUMBER_OF_FRAME` 로직을 조금 더 객체지향적으로 개선하려면 어떻게 해야할까요??

OhSeungCheol replied:
룰을 기준으로 마지막 프레임인지 판단하는 책임을 가지는 객체를 변경하였습니다 (BowlingGame -> Records)
[78a66fa](https://github.com/next-step/java-bowling/pull/903/commits/78a66fadcceb5fff234eba790bacc7ac315bfec6)

lxxjn0 commented on src/main/java/bowling/domain/FinalFrame.java:
```diff
@@ -2,15 +2,10 @@
 public class FinalFrame extends Frame {
     private boolean bonus;
-    private Scores bonusScores;
+    private Scores bonusScores = new Scores();
```
기존에 생성자에서 초기화하던 부분을 밖으로 뺀 이유가 있을까요?

OhSeungCheol replied:
New 키워드만 보고 빈 컬렉션이라고 생각했던 것 같습니다.  😅
생성자에서 초기화하도록 다시 변경했습니다.
[fadb0cf](https://github.com/next-step/java-bowling/pull/903/commits/fadb0cf730c836529e78424b5991b3c261b6d9dc)

lxxjn0 commented on src/main/java/bowling/domain/FinalFrame.java:
```diff
@@ -31,15 +26,25 @@ public void record(int downPinCount) {
     }
 
     public boolean validBonusScore() {
```
boolean을 반환하는 경우 is가 prefix로 붙으면 좀 더 가독성이 좋은 것 같아요!!

https://soojin.ro/blog/naming-boolean-variables

OhSeungCheol replied:
수정했습니다.
[77c8dcf](https://github.com/next-step/java-bowling/pull/903/commits/77c8dcf759b893f3725f80024f23dd9e8d9f1bfc)

lxxjn0 commented on src/main/java/bowling/domain/FinalFrame.java:
```diff
@@ -31,15 +26,25 @@ public void record(int downPinCount) {
     }
 
+    public int getSecondPitchScore() {
+        if (FrameScore.STRIKE.equals(getResult())) {
```
NPE 방지 👍🏼

lxxjn0 commented on src/main/java/bowling/view/ResultView.java:
```diff
@@ -9,24 +10,51 @@ public class ResultView {
+        try {
+            printTotalScores(game.getRecords());
+        } catch (NotReadyException ignored) {
+        }
```
예외를 catch하고 아무런 처리를 안하고 있는 것 같아요!!

OhSeungCheol replied:
Depth 제한이 있는 상태에서 for 루프를 break 할 방법이 생각이 안나서 시도한 방법인데요. 점수를 출력하다가 점수 준비가 안되어있으면 루프 종료의 기능만을 위한, 후 처리가 필요하지는 않은 예외라 무시했습니다.
for 기본 문법을 이용해서 개선했습니다.
[6e35dd1](https://github.com/next-step/java-bowling/pull/903/commits/6e35dd17e5c525f8455260dac6374769301809de)
[02f9bcd](https://github.com/next-step/java-bowling/pull/903/commits/02f9bcd549556bb26ebe4cd6842494f0a974e104)

lxxjn0 commented on src/test/java/bowling/domain/RecordsTest.java:
```diff
@@ -61,4 +63,322 @@ void FrameClassTest2() {
+        assertThat(result2).isEqualTo(FrameScore.STRIKE);
+        FrameScore result3 = records.getFrames().get(2).getResult();
+        assertThat(result3).isEqualTo(FrameScore.STRIKE);
+        assertThat(records.getTotalScore(0)).isEqualTo(10 + 10 + 10);
```
assertAll 메서드를 사용하면 어떤 부분이 다른지 학습해보고 활용해보면 좋을 것 같아요!!

OhSeungCheol replied:
개선했습니다.
[7672bbe](https://github.com/next-step/java-bowling/pull/903/commits/7672bbe5e38b4c8bfdcbaae82d127a17ded69a11)

lxxjn0 commented on src/main/java/bowling/view/ScoreString.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.util.Arrays;
+
+public enum ScoreString {
```
String과 같은 불용어를 네이밍에 사용하면 좋지 않아요. 이후 자료형의 변경과 같은 부분에서 자유로울 수 없기 때문입니다!! 출력과 관련된 다른 네이밍을 가져가는 것도 좋을 것 같아요!! 👍🏼

OhSeungCheol replied:
개선했습니다.
[01c11f6](https://github.com/next-step/java-bowling/pull/903/commits/01c11f690e60c91dbde4b93fbf5a5b23f4ca7753)

lxxjn0 commented on src/main/java/bowling/view/ScoreString.java:
```diff
@@ -0,0 +1,37 @@
+        return Arrays.stream(values())
+                .filter(vo -> vo.getFrameScore().equals(frameScore))
+                .findFirst()
+                .orElseThrow(RuntimeException::new)
```
적절한 예외 메시지도 함께 던져준다면 좋을 것 같아요!!

OhSeungCheol replied:
개선했습니다.
[16c46a3](https://github.com/next-step/java-bowling/pull/903/commits/16c46a3ef0703bdbe5d871b04a97d7813b361053)

lxxjn0 commented on src/main/java/bowling/view/ResultView.java:
```diff
@@ -9,24 +10,51 @@ public class ResultView {
     private static void printBonusScore(Frame frame) {
-        if (frame.validBonusScore()) {
-            FinalFrame finalFrame = (FinalFrame) frame;
+        if (!frame.getClass().equals(FinalFrame.class)) {
```
어떠한 클래스인지에 따라서 로직을 전개하는 것은 객체지향적인 관점에서 벗어나는 설계일 수 있어요. 아래의 글에서 해당 부분에 대한 설명이 자세히 나와있으니 한 번 참고해보시면 좋을 것 같아요!!

https://tecoble.techcourse.co.kr/post/2021-04-26-instanceof/

OhSeungCheol replied:
상속 관계 클래스별 분기 로직을 클래스 내부로 옮겨 개선했습니다.
[0e3e21d](https://github.com/next-step/java-bowling/pull/903/commits/0e3e21d3feed881d3e62777c30bb8a1e14c25372)

lxxjn0 reviewed (changes requested):
승철님, 미션 구현하시느라 수고 많으셨습니다!!

우선 간단한 부분에 대해서 피드백 남겨두었는데 한 번 확인해보시고 반영해주세요  👍🏼

Commit:
[M3] refactor: 마지막 프레임 판단 로직 캡슐화

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] refactor: 함수 네이밍 변경 (valid* -> isValid*)

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] feat: 프레임의 점수가 준비되었는지 판단하는 함수 추가

스트라이크이면서 다음 2 투구의 점수가 존재하거나,
스페어이면서 다음 1 투구의 점수가 존재하거나,
프레임이 종료되었는지를 검증

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] refactor: 불필요 try-catch 문 삭제

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] refactor: 불필요 멤버 변수 삭제

FinalFrame 의 보너스 게임 여부를 의미하는 bonus 변수를 삭제하고
프레임의 현재 상태를 기반으로 보너스 상태를 판단하는 함수를 활용

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] refactor: 프레임의 결과 판단 로직 함수화

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] refactor: 함수 네이밍 변경 (tryOver -> isTryOver)

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] test: 여러개의 상태를 테스트 할 경우 assertAll 을 활용하도록 수정

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] refactor: 클래스 네이밍 변경 (ScoreString -> ScorePhrase)

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] refactor: 예외 문구 추가

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] refactor: 함수 네이밍 변경 (isValidBonusScore -> isValidBonusGameScore)

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] refactor: 상속 관계 클래스 별 분기 로직 구조 변경

기존: 보너스 게임, 보너스 점수의 유효성을 Frame, FinalFrame 클래스 인지를 기준으로 외부에서 판단
수정: 보너스 게임, 보너스 점수의 유효성을 Frame 클래스가 직접 판단

참고:
https://github.com/next-step/java-bowling/pull/903
https://tecoble.techcourse.co.kr/post/2021-04-26-instanceof/

Commit:
[M3] refactor: 로직 단순화

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] fix: AllStrike 로 게임을 끝낼 경우 마지막 점수가 출력되지 않는 이슈 수정

9, 10(보너스 포함) 프레임이 전부 스트라이크일 때 9 프레임의 점수를 계산하는
과정에서 9 프레임의 다음 2번째 투구 보너스 점수를 11 프레임에서 찾으면서 예외 발생.
-> FinalFrame 의 보너스 스코어에서 점수를 조회하도록 수정

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] refactor: FinalFrame 생성자 추가

참고: https://github.com/next-step/java-bowling/pull/903

OhSeungCheol commented:
안녕하세요. 빠른 리뷰 감사합니다.

일부 내용 보완하여 커밋하였으니, 변경 부분에 대해 리뷰 한번 더 부탁드립니다.

감사합니다.

lxxjn0 commented on src/main/java/bowling/domain/FinalFrame.java:
```diff
@@ -37,14 +35,23 @@ public int getPinScore() {
+        return scores.getTryCount() + bonusScores.getTryCount();
+    }
+
+    public boolean isValidBonusScore() {
```
오버라이딩하는 메서드는 `@Override` 어노테이션을 붙여주면 좋을 것 같아요!!

OhSeungCheol replied:
추가했습니다.
[eee9d72](https://github.com/next-step/java-bowling/pull/903/commits/eee9d729e3eda9243ecab4a137aa4b71908cb581)

lxxjn0 commented on src/main/java/bowling/domain/Records.java:
```diff
@@ -38,6 +42,43 @@ public int getTotalScore(int targetIndex) {
         return totalScore;
     }
 
+    public boolean isReadyFrameScore(int frameIndex) {
```
해당 메서드도 메서드 분리를 한번 더 해주면 가독성이 좋아질 수 있을 것 같아요!!

OhSeungCheol replied:
수정했습니다.
[465c180](https://github.com/next-step/java-bowling/pull/903/commits/465c18075f0b1f717d37bcc887d6c9bdd33b8c91)

lxxjn0 reviewed (changes requested):
승철님, 피드백 반영해주시느라 수고하셨습니다!!

전체적으로 잘 개선해주셔서 크게 피드백을 드릴 부분이 없네요. 간단하게 한두개 피드백 남겨두었는데 확인 부탁드릴게요 👍🏼

Commit:
[M3] refactor: isReadyFrameScore 메소드 분리

isReadyBonusScore 메소드 추가

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] refactor: 중복 로직 함수화

isValidFrameIndex 메소드 추가

참고: https://github.com/next-step/java-bowling/pull/903

Commit:
[M3] refactor: Override 함수에 어노테이션 추가

참고: https://github.com/next-step/java-bowling/pull/903

OhSeungCheol commented:
안녕하세요. 리뷰 감사합니다.

상기 내용 보완했습니다!

감사합니다.

lxxjn0 reviewed (approved):
승철님, 확인이 늦어져서 죄송합니다 🙇🏼‍♂️ 

다음 단계 진행해주시면 될 것 같아요!! 👍🏼

OhSeungCheol commented:
감사합니다!
